pos_payment_terminal/static/src/js/devices.js

```diff
@@ -0,0 +1,114 @@
+                for (var transaction_id in response.transactions){
+                    var transaction = response.transactions[transaction_id];
+                    if ('transaction_id' in transaction){
+                        var transaction_id = transaction['transaction_id'];
+                        var line = order.get_payment_line_by_transaction(transaction_id);
+                        if (line === undefined){
+                            continue;
+                        }
+                        if ('success' in transaction){
+                            line.terminal_transaction_success = transaction['success'];
+                        }
+                        if ('status' in transaction){
+                            line.terminal_transaction_status = transaction['status'];
+                        }
+                        if ('reference' in transaction){
+                            line.terminal_transaction_reference = transaction['reference'];
+                        }
+                        if (line){
+                            paymentwidget.transaction_changed(line);
+                        }
+                    }
+                }
+            }
+        },
+        update_transaction_data: function(line, data){
+            data.amount = line.get_amount();
+            data.payment_mode = line.cashregister.journal.payment_mode;
+        },
+        handle_terminal_response: function(line_id, response){
+            // This is intended to entrich payment line with
+            // useful values (status, transaction_id, ...)
+            // Note that transaction_id is mandatory to match response
+            if (response === undefined){
+                return;
+            }
+            var self = this;
+            var order = self.pos.get_order()
+            var line = order.get_paymentline(line_id)
+            var paymentwidget = self.pos.chrome.screens.payment;
+            if ('transaction_id' in response){
+                line.terminal_transaction_id = response['transaction_id']
+            }
+            if ('success' in response){
+                line.terminal_success = response['success']
+            }
+            if ('status' in response){
+                line.terminal_status = response['status']
+            }
+            if ('reference' in response){
+                line.terminal_reference = response['reference']
+            }
+            paymentwidget.transaction_changed(line);
+        },
+        payment_terminal_transaction_start: function(line_cid, currency_iso, currency_decimals){
+            var self = this;
+            var order = this.pos.get_order();
+            var line = order.get_paymentline(line_cid);
+            var data = {
+                 'currency_iso' : currency_iso,
+                 'currency_decimals' : currency_decimals,
+                 'order_id': order.uid,
+            }
+            self.update_transaction_data(line, data)
```

Comment:
either inline `update_transaction_data` or make a prepare method for the whole `data` object